### PR TITLE
Rules: A function to transform a complex RSE expression into single-RSE expressions #4660

### PR DESCRIPTION
**Pull request by yiiyama**
Cutting out a functionality in transmogrified into an independent function in rse_selector.py.
This commit is still a draft and is open for discussions, including on which file to put this new function in.

### CODE REVIEW

File: lib/rucio/core/rse_selector.py [hunk 1]
```diff
@@ -247,3 +248,34 @@ def __choose_rse(self, rses):
...
+
+    
+@read_session
+def resolve_rse_expression(rse_expression, account, weight=None, copies=1, ignore_account_limit=False, size=0, preferred_rses=[], blocklist=[], prioritize_order_over_weight=False, existing_rse_size=None, session=None):
+    """
+    Resolve a potentially complex RSE expression into `copies` single-RSE expressions. Uses `parse_expression()`
+    to decompose the expression, then `RSESelector.select_rse()` to pick the target RSEs.
+    """
+    
+    rses = parse_expression(rse_expression, filter_={'vo': account.vo})
```
Comment (bari12) [inline lib/rucio/core/rse_selector.py]: You need to pass in the `session` here, otherwise this is a leak.

File: lib/rucio/core/rse_selector.py [hunk 2]
```diff
@@ -247,3 +248,34 @@ def __choose_rse(self, rses):
...
+
+    rse_to_id = dict((rse_dict['rse'], rse_dict['id']) for rse_dict in rses)
+    id_to_rse = dict((rse_dict['id'], rse_dict['rse']) for rse_dict in rses)
+    
+    selector = RSESelector(account=account,
+                           rses=rses,
+                           weight=weight,
+                           copies=copies,
+                           ignore_account_limit=ignore_account_limit)
```
Comment (bari12) [inline lib/rucio/core/rse_selector.py]: Session leak

**Review by bari12 — DISMISSED** [review]
Two session leaks which you need to pass through. In principle this is GTG though. Maybe @cserf should have a look as well on the transmo part!

**Comment by yiiyama** [comment]
Thank you @bari12 for the review. I'm sorry it took more than 10 days to make the fixes. I also noticed that I had made a stupid typo in the newly added import statement. It's fixed now.

**Review by cserf — APPROVED** [review]
LGTM

**Pull request merged**